mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -1017,6 +1027,100 @@ def _search_datasets(self, experiment_ids):
+        """
+        Creates a run for prompt engineering with the specified attributes.
+        """
+        run = self.create_run(experiment_id, user_id, start_time, tags, run_name)
+        run_id = run.info.run_id
+
+        # log model parameters
+        parameters_to_log = [
+            *model_parameters,
+            Param("model_route", model_route),
+            Param("prompt_template", prompt_template),
+        ]
+
+        self.log_batch(run_id, [], parameters_to_log, [])
+
+        # set logged artifacts tag
+        tag_value = [{"path": "eval_results_table.json", "type": "table"}]
+        self.set_tag(run_id, RunTag(MLFLOW_LOGGED_ARTIFACTS, json.dumps(tag_value)))
+
+        # set the promptlab run tag
+        self.set_tag(run_id, RunTag(MLFLOW_RUN_SOURCE_TYPE, "PROMPT_ENGINEERING"))
+
+        artifact_dir = self._get_artifact_location(experiment_id)
+        artifact_repo = get_artifact_repository(artifact_dir)
+
+        # log model
+        from mlflow.models import Model
+
+        utc_time_created = datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S.%f")
+        promptlab_model = Model(
+            artifact_path=artifact_dir,
+            run_id=run_id,
+            utc_time_created=utc_time_created,
+        )
+        self.record_logged_model(run_id, promptlab_model)
+
+        try:
+            from mlflow.models.signature import ModelSignature
+            from mlflow.types.schema import ColSpec, DataType, Schema
+
+            inputs_colspecs = [ColSpec(DataType.string, param.key) for param in prompt_parameters]
+            outputs_colspecs = [ColSpec(DataType.string, "output")]
+            signature = ModelSignature(
+                inputs=Schema(inputs_colspecs),
+                outputs=Schema(outputs_colspecs),
+            )
+        except ImportError:
+            signature = None
+
+        from mlflow._promptlab import save_model
+
+        # write artifact files
+        with tempfile.TemporaryDirectory() as local_dir:
+            save_model(
+                path=os.path.join(local_dir, "model"),
+                signature=signature,
+                input_example={"inputs": [param.value for param in prompt_parameters]},
+                prompt_template=prompt_template,
+                prompt_parameters=prompt_parameters,
+                model_parameters=model_parameters,
+                model_route=model_route,
+            )
+
+            eval_results_json = create_eval_results_file(
+                prompt_parameters, model_input, model_output_parameters, model_output
+            )
+            eval_results_json_file_path = os.path.join(local_dir, "eval_results_table.json")
+            make_containing_dirs(eval_results_json_file_path)
+            write_to(eval_results_json_file_path, eval_results_json)
+
+            artifact_repo.log_artifacts(local_dir)
+
+        self.update_run_info(
+            run_id, RunStatus.FINISHED, int(datetime.now().strftime("%Y%m%d")), run_name
+        )
```

Comment:
Is this line correct? `int(datetime.now().strftime("%Y%m%d"))` returns an integer like `20230831`.

Reply:
good catch. this should be `int(time.time() * 1000)`.

on this the RunInfo docs https://mlflow.org/docs/latest/python_api/mlflow.entities.html?highlight=end_time#mlflow.entities.RunInfo.end_time
> End time of the run, in number of milliseconds since the UNIX epoch.